indexer/indexer.go

```diff
@@ -849,7 +1015,43 @@ func (collection *Collection) Search(ctx context.Context, query Query) (<-chan R
 				}
 			}
 		}(c)
+	}
 
+	if query.Sort != "" {
```

Comment:
Sort should be done as the index is being traversed, not after the fact.

Note the logic here where we choose the best index that works with both $eq and sort: https://github.com/RangerMauve/hyperbeedeebee/blob/default/index.js#L356

Here is how the start key gets created for the index which takes sort into consideration: https://github.com/RangerMauve/hyperbeedeebee/blob/default/index.js#L482

We also change the direction we iterate based on the sort parameter:

https://github.com/RangerMauve/hyperbeedeebee/blob/default/index.js#L485

Reply:
Sry, I'm pool in javascript, can you convert the codes to algorithm descriptions?

Reply:
Inside BestIndex we need the following added around [here](https://github.com/RangerMauve/ipld-prolly-indexer/blob/c8a3106813ab115332b6c2767e18bb96211b0168/indexer/indexer.go#L1086). This is based on the code [here](https://github.com/RangerMauve/hyperbeedeebee/blob/default/index.js#L356C20-L356C20)

If sort is specifid in the query:
- if the current index doesn't contain the sort field, skip it
- check that all the fields before the sort field are in the `Query.Equals` map. If there is a field before the Sort field that isn't in the Equals map, skip this index.
- If the number of consecutive fields > `bestMatchingFields`, set `best` to the index and `bestMatchingFields` to the consecutive fields

This will ensure we have the best index for the query to do sequential reads.

Next inside [this if statement](https://github.com/RangerMauve/ipld-prolly-indexer/blob/c8a3106813ab115332b6c2767e18bb96211b0168/indexer/indexer.go#L923) if sort is not null, return an error.

So with this we'll pick indexes where we can seek to Equals values then iterate as we're sorting.

I'm not sure whether we can iterate backwards in the prolly tree yet, but we might need to add that in.

Reply:
Sorry, I'm confused about the `before the sort field` and `consecutive fields`, can you elaborate it?

Reply:
Say we have a records that look like this:

```javascript
{
author,
createdAt,
type,
tag,
content
}
```

and we're doing a search that looks like

```
Equals: {
  author: "alice",
type: "note",
tag: "cats"
}
Sort: createdAt
```

Say we have the following indexes:

```
[createdAt]
[type, createdAt]
[author, type, createdAt]
[createdAt, type, author, tag]
```

For this query the third index would be best because we can have the start key set to `[alice, note,0x00]` and then all subsequent keys will be sorted by the `createdAt` field.

Notice that the fourth index has more fields from the query, but because they come after the `createdAt` field we can't use them as an easy filter by jumping in the keyspace and would need to iterate over all the documents in the index rather than just the ones with author: alice and type:note.

Does that help?

Reply:
I see, do you mean the best index for query must start from specific  and consecutive fields(longest mean the best index) and end with the sort field?

Reply:
But I'm a little confused
https://github.com/RangerMauve/ipld-prolly-indexer/blob/c8a3106813ab115332b6c2767e18bb96211b0168/indexer/indexer.go#L1091
The code here doesn't judge whether consecutive,  do we need modify it?

Reply:
> If here is missing field before sort field or not include the sort field we should skip it

Yes exactly. It's okay if there are other fields after the sort field, it's just that everything before the sort field must be inside Equals

> The code here doesn't judge whether consecutive, do we need modify it?

That code should be used for non-sorted index search. At that point just having the most fields indexed is the best option.

Reply:
Thanks, I have commited the new code, but I'm still confused about the index that if fields in equalmap are not consecutive in the best index,  such as.   equalmap{A:a, B:b, D:d}.  and bestindex is [A,B,C,D],   not sort,   how does it work?

Reply:
Yeah, the longest consecutive fields would still be best, but having non-consecutive is still useful because it allows for trying to fetch fields out of the index key without needing to load the full record. But tbh it might be less relevant to how our code is now. Do you think we should sort by "most consecutive" and add the "must end in sort field" logic if sort is specified?

Reply:
I think "most consecutive" for BestIndex is must but "must end in sort field" is not need.
https://github.com/RangerMauve/ipld-prolly-indexer/blob/86e709e4d895d6b514cb21097aab44370bbc17c8/indexer/indexer.go#L629
I'm confused about here only used the consecutive part